### ISSUE TITLE
Expressions for mean, stddev, etc., for circular values (e.g., angle)

### ISSUE THREAD

**Issue by warpedgeoid** [pola-rs/polars#11714]
### Description

It does not appear that Polars has built-in expressions for calculating statistics (e.g., mean, stddev) properly on circular data during aggregation. An example would be a column containing angle measurements where wrapping occurs at 360°. This is supported by many scientific computing libraries. Would it be possible to get circmean(), circstddev(), etc., expressions to prevent having to use a Python function during aggregation of very large (200M+ rows) datasets?

You can read more about this in the SciPy docs:
[https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.circmean.html](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.circmean.html)

**Comment by svaningelgem** [comment]
I doubt this should be implemented?
```python
import polars as pl
from scipy.stats import circmean
import numpy as np

angles = np.deg2rad(np.array([20, 30, 330]))

df = pl.DataFrame({
    "deg": angles,
})

print(np.rad2deg(circmean(df["deg"])))
# 7.294976657784009
```

**Comment by warpedgeoid** [comment]
I'm not sure how this example helps with the stated problem, which is to avoid calling the Python function when aggregating. Also, the dataframe in question is hundreds of millions of records and must be processed using the lazy API unless running on a HPC.

**Comment by warpedgeoid** [comment]
You're right that Pandas doesn't have native support either; however, Pandas is already really slow so you don't notice the slowdown 😁 

How difficult do you think it would be for someone with basic Rust skills who hasn't contributed to Polars before, to implement this functionality? Or, would it be possible for me to implement some sort of extension to Polars that provides the functionality only to those who actually need it?

**Comment by cmdlineluser** [comment]
Just looking at what `scipy` itself does: https://github.com/scipy/scipy/blob/v1.11.3/scipy/stats/_morestats.py#L4434
```python
import math
import scipy.stats
import polars as pl
import numpy as np

angles = np.deg2rad(np.array([20, 30, 330]))

df = pl.DataFrame({
    "deg": angles,
})

scipy.stats.circmean(df["deg"])
# 0.12732136153446258
```
It seems the basic functionality is essentially:
```python
def circmean(expr, high=math.pi * 2, low=0):
   sin_samp = (((expr - low) * 2 * math.pi) / (high - low)).sin()
   cos_samp = (((expr - low) * 2 * math.pi) / (high - low)).cos()
   
   return pl.arctan2(sin_samp.sum(), cos_samp.sum())
   
df.select(circmean(pl.col("deg")))
# shape: (1, 1)
# ┌──────────┐
# │ deg      │
# │ ---      │
# │ f64      │
# ╞══════════╡
# │ 0.127321 │
# └──────────┘
```
If you wanted to implement it at the Python level, you could perhaps register a namespace: https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.api.register_expr_namespace.html

Which would allow you to create `.circ.mean()` etc.

**Comment by svaningelgem** [comment]
> If you wanted to implement it at the Python level
That's exactly what @warpedgeoid doesn't want to do 😁. And no, it's not difficult to add.

@stinodego : would this be something that's good to add? (and if yes: @warpedgeoid , maybe make a list of functions to implement, ordered from most useful to nice to have?)

**Comment by MarcoGorelli** [comment]
> That's exactly what @warpedgeoid doesn't want to do

This wouldn't be Python though - it would use the Python API of Polars, but it would be Rust code underneath and so should be fast, and it should work with the lazy api, e.g.

```python
In [22]: df = pl.LazyFrame({'deg': [angles]})

In [23]: df.with_columns(circmean=pl.col('deg').list.eval(circmean(pl.element()))).collect()
Out[23]:
shape: (1, 2)
┌────────────────────────────────┬────────────┐
│ deg                            ┆ circmean   │
│ ---                            ┆ ---        │
│ list[f64]                      ┆ list[f64]  │
╞════════════════════════════════╪════════════╡
│ [0.349066, 0.523599, 5.759587] ┆ [0.127321] │
└────────────────────────────────┴────────────┘
```

**Comment by svaningelgem** [comment]
> If you did want to do this purely in Rust, you may want to take a look at https://github.com/pola-rs/pyo3-polars

I'm definitly going to do that!

**Comment by warpedgeoid** [comment]
> > That's exactly what @warpedgeoid doesn't want to do
> 
> This wouldn't be Python though - it would use the Python API of Polars, but it would be Rust code underneath and so should be fast, and it should work with the lazy api
> 
> It should also work within a group by context, which I think is what @warpedgeoid is after
> 
> quick timing test:
> 
> ```python
> In [1]: import math
> 
> In [2]: def circmean(expr, high=math.pi * 2, low=0):
>    ...:    sin_samp = (((expr - low) * 2 * math.pi) / (high - low)).sin()
>    ...:    cos_samp = (((expr - low) * 2 * math.pi) / (high - low)).cos()
>    ...:
>    ...:    return pl.arctan2(sin_samp.sum(), cos_samp.sum())
>    ...:
> 
> In [3]: arr = np.deg2rad(np.random.randn(20_000_000))
> 
> In [4]: df = pl.DataFrame({'deg': arr})
> 
> In [5]: import scipy.stats
> 
> In [6]: %timeit df.select(circmean(pl.col("deg")))
> 299 ms ± 21.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
> In [7]: %timeit scipy.stats.circmean(df["deg"])
> 515 ms ± 21.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```
> 
> If you did want to do this purely in Rust, you may want to take a look at https://github.com/pola-rs/pyo3-polars

I'll admit to not having thought about a function that builds a more complex expression. I also didn't realize that all of the trig functions were implemented in Polars. Let me test this on my real data to compare performance and results.

Also, thanks for the link to pyo3-polars. I'm actually using pyo3 in the current project to implement file scanning and analysis modules in Rust for use in Python notebooks.

**Comment by deanm0000** [comment]
> If you wanted to implement it at the Python level, you could perhaps register a namespace: https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.api.register_expr_namespace.html

You could also just monkey patch it to `pl.Expr` ie.

`pl.Expr.circmean=circmean` which then allows

`df.select(pl.col('deg').circmean())`

As a bit of a tangent, the whole (or at least huge majority) of [scipy.special](https://docs.scipy.org/doc/scipy/reference/special.html) functions are ufuncs that you can use directly in expressions as detailed [here](https://pola-rs.github.io/polars/user-guide/expressions/numpy/) Many functions in scipy.stats are really just aliases (is that the right term) of the ufunc in scipy.special.

**Comment by warpedgeoid** [comment]
> > If you wanted to implement it at the Python level, you could perhaps register a namespace: https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.api.register_expr_namespace.html
> 
> 
> 
> You could also just monkey patch it to `pl.Expr` ie.
> 
> 
> 
> `pl.Expr.circmean=circmean` which then allows
> 
> 
> 
> `df.select(pl.col('deg').circmean())`
> 

I'm not sure I could live with myself 😂

> 
> 
> As a bit of a tangent, the whole (or at least huge majority) of [scipy.special](https://docs.scipy.org/doc/scipy/reference/special.html) functions are ufuncs that you can use directly in expressions as detailed [here](https://pola-rs.github.io/polars/user-guide/expressions/numpy/) Many functions in scipy.stats are really just aliases (is that the right term) of the ufunc in scipy.special.

This is interesting, but it seems like it might have trouble with missing values.

**Comment by alexander-beedie** [comment]
> > If you wanted to implement it at the Python level
>
> That's exactly what @warpedgeoid doesn't want to do 😁. And no, it's not difficult to add.

Instead of "implement", read "expose"; 100% of the calculation would be in native Rust using this approach ;)

**Comment by stinodego** [comment]
We do not plan on implementing this as it is too niche, and it can be implemented using helper functions using existing expressions.

This might be a good candidate for a cookbook: #13064

**Comment by warpedgeoid** [comment]
Can confirm that the provided workaround is sufficient. Thanks!
